File: machine-learning/small-tasks/cross-validation.cpp
#include <iostream>
#include <algorithm>
#include <numeric>
#include <vector>

int main() {
    size_t object_amount, classes_amount, groups_amount;
    std::cin >> object_amount >> classes_amount >> groups_amount;

    size_t objects[object_amount];
    size_t class_object_amount[classes_amount + 1] = {};

    for (size_t i = 0; i < object_amount; ++i) {
        std::cin >> objects[i];
        ++class_object_amount[objects[i]];
    }
    std::partial_sum(class_object_amount, class_object_amount + classes_amount + 1, class_object_amount);
    std::vector<size_t> groups[groups_amount];

    for (size_t i = 0; i < object_amount; ++i) {
        size_t tmp = objects[i] - 1;
        size_t& cl = class_object_amount[tmp];
        groups[cl % groups_amount].push_back(i + 1);
        ++cl;
    }

    std::for_each(groups, groups + groups_amount, [] (auto x) {
        std::cout << x.size() << ' ';
        std::for_each(x.begin(), x.end(), [] (auto y) { std::cout << y << ' '; });
        std::cout << '\n';
    });

    return 0;
}

File: machine-learning/small-tasks/cross-validation.py
from itertools import accumulate

read_ints = lambda: list(map(int, input().split()))

objects, classes, groups = read_ints()
cl = [0] * classes
ob = read_ints()

for i in ob:
    cl[i - 1] += 1

cl = list(accumulate([0] + cl))
gr = [[] for x in range(groups)]

for i in range(len(ob)):
    tmp = ob[i] - 1
    c = cl[tmp]
    gr[c % groups].append(i + 1)
    cl[tmp] += 1

for i in gr:
    print(len(i), end=' ')
    for j in i:
        print(j, end=' ')
    print()
